iothub.class.nut

```diff
@@ -5,724 +5,1024 @@
-        _transport.sendFeedback(action, message, done);
-        return this;
-    }
+            return this;
+        }
 
-}
+        function remove(deviceId = null, done = null) {
 
-//------------------------------------------------------------------------------
-class iothub.RegistryHTTP {
+            // NOTE: These default values are not from the original Node.js SDK
+            if (typeof deviceId == "function") {
+                done = deviceId;
+                deviceId = null;
+            }
+            if (typeof deviceId != "string") {
+                deviceId = split(http.agenturl(), "/").pop();
+            }
 
-    _config = null;
+            local path = iothub.Endpoint.devicePath(deviceId) + iothub.Endpoint.versionQueryString();
+            _transport.deleteDevice(path, done);
 
-    constructor(config) {
-        _config = config;
+            return this;
+        }
     }
 
-    function refreshSignature() {
+    //------------------------------------------------------------------------------
+
+    // Client transport class
+    // NOTE: This class differs from the original Node.js SDK
+    ClientAMQP = class {
+
+        _config = null;
+        _connection = null;
+        _sessions = null;
+        _senders = null;
+        _receivers = null;
+        _transfers = null;
+        _handlers = null;
+        _msgQueue = null;
+
+        _connecting = false;
+        _senderTokenError = false;
+        _receiverTokenError = false;
+        _debug = false;
+
+        constructor(config) {
+            _config = config;
+            _resetConnectionTables();
+        }
 
-        // NOTE: This method did not appear in the original Node.js SDK
-        if ("sharedAccessExpiry" in _config && "connectionString" in _config) {
-            if (time() >= _config.sharedAccessExpiry) {
-                local cn = iothub.ConnectionString.Parse(_config.connectionString);
-                local sas = iothub.SharedAccessSignature.create(cn.HostName, cn.SharedAccessKeyName, cn.SharedAccessKey, iothub.Authorization.anHourFromNow());
-                _config.sharedAccessSignature = sas.toString();
-                _config.sharedAccessExpiry = sas.se;
-                // server.log("Signature refreshed");
+        // done has one param (error)
+        function connect(done = null) {
+            // set connecting flag
+            _connecting = true;
+            // set connection callback
+            _handlers.onConnected <- done;
+
+            // Don't open a connection if one is already open
+            if( _isOpen(_connection) ) {
+                if ( _isOpen(_sessions.auth) ) {
+                    if ( _isOpen(_receivers.auth) ) {
+                        if ( _isOpen(_senders.auth) ) {
+                            if ( _isOpen(_sessions.event) ) {
+                                if ( _isOpen(_senders.event) ) {
+                                    _connectionCallback(null);
```

Comment:
ooh! I feel like this should have been initially implemented as a finite state machine which could be way much cleaner and transparent. Now I feel like we'll have a hard time to maintain all these dependencies :) but this is too late to rewrite everything. So I'm just thinking loudly...